### PR TITLE
do not print wrapping ranges like normal ranges in validity diagnostics

### PR DESCRIPTION
**Pull request by RalfJung**

### CODE REVIEW

File: src/librustc_mir/interpret/validity.rs [hunk 1]
```diff
@@ -140,6 +141,33 @@ fn scalar_format<Tag>(value: ScalarMaybeUndef<Tag>) -> String {
...
+fn wrapping_range_format(r: &RangeInclusive<u128>, max_hi: u128) -> String {
+    let (lo, hi) = r.clone().into_inner();
+    debug_assert!(hi <= max_hi);
+    if lo > hi {
+        format!("less or equal to {}, or greater or equal to {}", hi, lo)
+    } else {
+        if hi == max_hi {
+            format!("greater or equal to {}", lo)
+        } else {
+            format!("in the range {:?}", r)
```
Comment (oli-obk) [inline src/librustc_mir/interpret/validity.rs]: all the `IndexVec` indices are `smaller or equal to`. Seems like something reasonable to support, too?

File: src/librustc_mir/interpret/validity.rs [hunk 2]
```diff
@@ -140,6 +141,33 @@ fn scalar_format<Tag>(value: ScalarMaybeUndef<Tag>) -> String {
...
+    }
+}
+
+// Formats such that a sentence like "expected something {}" to mean
+// "expected something <in the given range>" makes sense.
+fn wrapping_range_format(r: &RangeInclusive<u128>, max_hi: u128) -> String {
+    let (lo, hi) = r.clone().into_inner();
+    debug_assert!(hi <= max_hi);
+    if lo > hi {
+        format!("less or equal to {}, or greater or equal to {}", hi, lo)
```
Comment (oli-obk) [inline src/librustc_mir/interpret/validity.rs]: There are no tests that show this message

Comment (RalfJung) [reply src/librustc_mir/interpret/validity.rs]: Any suggestions for how to make one?

File: src/librustc_mir/interpret/validity.rs (hunk 1 above)
Comment (RalfJung) [reply src/librustc_mir/interpret/validity.rs]: I can add the printing, but how would I go about testing it?

Comment (RalfJung) [reply src/librustc_mir/interpret/validity.rs]: Oh actually we have tons of tests for this, from `NonZero`.

File: src/librustc_mir/interpret/validity.rs (hunk 2 above)
Comment (oli-obk) [reply src/librustc_mir/interpret/validity.rs]:
```rust
#![feature(rustc_attrs)]
#![feature(const_transmute)]

#[rustc_layout_scalar_valid_range_start(10)]
#[rustc_layout_scalar_valid_range_end(30)]
struct Foo(u32);

const F: Foo = unsafe { Foo(42) };
```

Comment (RalfJung) [reply src/librustc_mir/interpret/validity.rs]: That will show `in the range 10..=30`, which we probably also didn't test before.

But if I swap the two ranges, I get the error you asked for.

**Pull request merged**